### PR TITLE
Improve the guest OS hypervisor mappings addition on upgrade.

### LINKED ISSUE #5685: duplicate entry in guest_os when upgrading from 4.14 to 4.15
##### ISSUE TYPE

 * Bug Report

##### COMPONENT NAME

~~~
DB
~~~

##### CLOUDSTACK VERSION

~~~
4.14.1 => 4.15.2
~~~

##### CONFIGURATION

- existing entry in guest_os with id 305
- upgrading vom 4.14.1 to 4.15.2

##### OS / ENVIRONMENT

N/A

##### SUMMARY

When upgrading from 4.14.1 to 4.15.2 and from whatever reason a entry with id 305 in guest os exist, the db schema upgrading script fails

##### STEPS TO REPRODUCE

~~~
restarting the management server after the packet upgrade fails with error "duplicate entry with id 305...". removing the entry and setting autoincrement back to 305 allows a successful migration

mysql> select * from guest_os_hypervisor where guest_os_id = 305;
+-------+-----------------+---------------------+-------------+--------------------+--------------------------------------+---------------------+---------+-----------------+
| id    | hypervisor_type | guest_os_name       | guest_os_id | hypervisor_version | uuid                                 | created             | removed | is_user_defined |
+-------+-----------------+---------------------+-------------+--------------------+--------------------------------------+---------------------+---------+-----------------+
| 12345 | VMware          | other3xLinux64Guest |         305 | 6.5                | xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx | 2020-08-22 19:44:35 | NULL    |               1 |
+-------+-----------------+---------------------+-------------+--------------------+--------------------------------------+---------------------+---------+-----------------+

mysql> select * from guest_os where id = 305;
+-----+-------------+------+--------------------------------------+--------------------------+---------------------+---------+-----------------+
| id  | category_id | name | uuid                                 | display_name             | created             | removed | is_user_defined |
+-----+-------------+------+--------------------------------------+--------------------------+---------------------+---------+-----------------+
| 305 |           7 | NULL | xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx | Other 3.x Linux (64-bit) | 2020-08-22 21:43:15 | NULL    |               1 |
+-----+-------------+------+--------------------------------------+--------------------------+---------------------+---------+-----------------+

mysql> delete from guest_os_hypervisor where id = 12345;
mysql> delete from guest_os where id = 305;
mysql> ALTER TABLE guest_os AUTO_INCREMENT=305;

~~~

##### EXPECTED RESULTS

newly created entry should not have an id set, it should use autoincrement

~~~
newly created entry should not have an id set, it should use autoincrement
~~~

##### ACTUAL RESULTS

~~~
script fails with sql error concerning duplicate entry with id 305
~~~

### PR DESCRIPTION
**Pull request by sureshanaparti**
### Description

This PR improves the guest OS hypervisor mappings addition on upgrade. Also, removes the hardcoded guest os ids 305 - 329 in upgrade path 4.14.0 to 4.15.0, and 330 - 349 in upgrade path 4.15.0.0 to 4.15.1.0.

Please note, the hypervisor mapping for the guest OSes added pre-4.14 (with hardcoded ids < 305), are updated using the same guest os ids.

Fixes #5685

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [x] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by sureshanaparti** [comment]
@blueorangutan package

File: engine/schema/src/main/java/com/cloud/upgrade/dao/Upgrade41510to41520.java [hunk 1]
```diff
@@ -51,6 +58,12 @@ public boolean supportsRollingUpgrade() {
 
     @Override
     public void performDataMigration(Connection conn) {
+        correctGuestOsIdsInHypervisorMapping(conn);
+    }
+
+    private void correctGuestOsIdsInHypervisorMapping(final Connection conn) {
+        LOG.debug("Correcting guest OS ids in hypervisor mappings");
+        guestOsMapper.updateGuestOsIdInHypervisorMapping(conn, 10, "Ubuntu 20.04 LTS", new Ternary<String, String, String>("Xenserver", "8.2.0", "Ubuntu Focal Fossa 20.04"));
```
Comment (sureshanaparti) [inline engine/schema/src/main/java/com/cloud/upgrade/dao/Upgrade41510to41520.java:66]: This would fix the incorrect guest os mapping in 4.15.0 to 4.15.1 schema, when upgrading to 4.15.2.

https://github.com/apache/cloudstack/blob/571339f8724fc70c472f55f254125f0c6070d5de/engine/schema/src/main/resources/META-INF/db/schema-41500to41510.sql#L55-L57

File: engine/schema/src/main/java/com/cloud/upgrade/dao/Upgrade41520to41600.java [hunk 2]
```diff
@@ -71,6 +74,12 @@ public boolean supportsRollingUpgrade() {
     public void performDataMigration(Connection conn) {
         generateUuidForExistingSshKeyPairs(conn);
         populateAnnotationPermissions(conn);
+        correctGuestOsIdsInHypervisorMapping(conn);
+    }
+
+    private void correctGuestOsIdsInHypervisorMapping(final Connection conn) {
+        LOG.debug("Correcting guest OS ids in hypervisor mappings");
+        guestOsMapper.updateGuestOsIdInHypervisorMapping(conn, 10, "Ubuntu 20.04 LTS", new Ternary<String, String, String>("Xenserver", "8.2.0", "Ubuntu Focal Fossa 20.04"));
```
Comment (sureshanaparti) [inline engine/schema/src/main/java/com/cloud/upgrade/dao/Upgrade41520to41600.java:82]: This would fix the incorrect guest os mapping in 4.15.0 to 4.15.1 schema, when upgrading to 4.16.0 from existing 4.15.2 env.

https://github.com/apache/cloudstack/blob/571339f8724fc70c472f55f254125f0c6070d5de/engine/schema/src/main/resources/META-INF/db/schema-41500to41510.sql#L55-L57

File: engine/schema/src/main/java/com/cloud/upgrade/dao/Upgrade41600to41610.java [hunk 3]
```diff
@@ -57,6 +60,12 @@ public boolean supportsRollingUpgrade() {
 
     @Override
     public void performDataMigration(Connection conn) {
+        correctGuestOsIdsInHypervisorMapping(conn);
+    }
+
+    private void correctGuestOsIdsInHypervisorMapping(final Connection conn) {
+        LOG.debug("Correcting guest OS ids in hypervisor mappings");
+        guestOsMapper.updateGuestOsIdInHypervisorMapping(conn, 10, "Ubuntu 20.04 LTS", new Ternary<String, String, String>("Xenserver", "8.2.0", "Ubuntu Focal Fossa 20.04"));
```
Comment (sureshanaparti) [inline engine/schema/src/main/java/com/cloud/upgrade/dao/Upgrade41600to41610.java:68]: This would fix the incorrect guest os mapping in 4.15.0 to 4.15.1 schema, when upgrading to 4.16.1 from existing 4.16.0 env.

https://github.com/apache/cloudstack/blob/571339f8724fc70c472f55f254125f0c6070d5de/engine/schema/src/main/resources/META-INF/db/schema-41500to41510.sql#L55-L57

File: engine/schema/src/main/java/com/cloud/upgrade/dao/Upgrade41500to41510.java [hunk 4]
```diff
@@ -235,6 +242,138 @@ public void updateSystemVmTemplates(final Connection conn) {
         LOG.debug("Updating System Vm Template IDs Complete");
     }
 
+    private void correctGuestOsNames(final Connection conn) {
```
Comment (sureshanaparti) [inline engine/schema/src/main/java/com/cloud/upgrade/dao/Upgrade41500to41510.java:245]: This corrects the guest os names added in 4.14.0 to 4.15.0 schema.

https://github.com/apache/cloudstack/blob/571339f8724fc70c472f55f254125f0c6070d5de/engine/schema/src/main/resources/META-INF/db/schema-41400to41500.sql#L729

https://github.com/apache/cloudstack/blob/571339f8724fc70c472f55f254125f0c6070d5de/engine/schema/src/main/resources/META-INF/db/schema-41400to41500.sql#L739

https://github.com/apache/cloudstack/blob/571339f8724fc70c472f55f254125f0c6070d5de/engine/schema/src/main/resources/META-INF/db/schema-41400to41500.sql#L805-L807

https://github.com/apache/cloudstack/blob/571339f8724fc70c472f55f254125f0c6070d5de/engine/schema/src/main/resources/META-INF/db/schema-41500to41510.sql#L21-L24

File: engine/schema/src/main/java/com/cloud/upgrade/dao/Upgrade41500to41510.java [hunk 5]
```diff
@@ -235,6 +242,138 @@ public void updateSystemVmTemplates(final Connection conn) {
...
+        mappings.add(new Ternary<String, String, String>("Xenserver", "8.1.0", "Scientific Linux 7"));
+        guestOsMapper.addGuestOsAndHypervisorMappings (conn, 9, "Scientific Linux 7", mappings);
+        mappings.clear();
+
+        // Add support for NeoKylin Linux Server 7 for Xenserver 8.1.0
+        guestOsMapper.addGuestOsAndHypervisorMappings(conn, 9, "NeoKylin Linux Server 7", mappings); //334
+        mappings.clear();
+
+        // Pass Guest OS Ids to update pre-4.14 mappings
+        // Add support CentOS 8 for Xenserver 8.1.0
```
Comment (sureshanaparti) [inline engine/schema/src/main/java/com/cloud/upgrade/dao/Upgrade41500to41510.java:283]: Please note, the hypervisor mapping for the guest OSes added pre-4.14 (with hardcoded ids < 305), are updated using the same ids.

File: engine/schema/src/main/java/com/cloud/upgrade/GuestOsMapper.java [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+    private static final String selectGuestOsSql =
+            "SELECT id FROM `cloud`.`guest_os` WHERE category_id = ? AND display_name = ? AND is_user_defined = 0 AND removed IS NULL ORDER BY created DESC";
+    private static final String insertGuestOsSql =
+            "INSERT INTO `cloud`.`guest_os` (uuid, category_id, display_name, created) VALUES (UUID(), ?, ?, now())";
+    private static final String selectGuestOsHypervisorSql =
+            "SELECT guest_os_id FROM `cloud`.`guest_os_hypervisor` WHERE hypervisor_type = ? AND hypervisor_version = ? AND guest_os_name = ? AND is_user_defined = 0 AND removed IS NULL ORDER BY created DESC";
+    private static final String insertGuestOsHypervisorSql =
+            "INSERT INTO `cloud`.`guest_os_hypervisor` (uuid, hypervisor_type, hypervisor_version, guest_os_name, guest_os_id, created) VALUES (UUID(), ?, ?, ?, ?, now())";
+    private static final String updateGuestOsSql =
+            "UPDATE `cloud`.`guest_os` SET display_name = ? WHERE id = ?";
```
Comment (DaanHoogland) [inline engine/schema/src/main/java/com/cloud/upgrade/GuestOsMapper.java:43]: is the `display_name` the only field we'd want to update?

**Comment by sureshanaparti** [comment]
@blueorangutan test keepEnv

File: engine/schema/src/main/resources/META-INF/db/schema-41500to41510.sql [hunk 7]
```diff
@@ -18,117 +18,20 @@
...
-INSERT INTO `cloud`.`guest_os_hypervisor` (uuid,hypervisor_type, hypervisor_version, guest_os_name, guest_os_id, created, is_user_defined) VALUES (UUID(),'VMware', '7.0.1.0', 'rhel9_64Guest', 348, now(), 0);
-
--- Add support for sles16_64Guest from VMware 7.0.1.0
-INSERT INTO `cloud`.`guest_os` (id, uuid, category_id, display_name, created) VALUES (349, UUID(), 5, 'SUSE Linux Enterprise Server 16 (64-bit)', now());
-INSERT INTO `cloud`.`guest_os_hypervisor` (uuid,hypervisor_type, hypervisor_version, guest_os_name, guest_os_id, created, is_user_defined) VALUES (UUID(),'VMware', '7.0.1.0', 'sles16_64Guest', 349, now(), 0);
-
--- Add support for windows2019srvNext_64Guest from VMware 7.0.1.0
-INSERT INTO `cloud`.`guest_os_hypervisor` (uuid,hypervisor_type, hypervisor_version, guest_os_name, guest_os_id, created, is_user_defined) VALUES (UUID(),'VMware', '7.0.1.0', 'windows2019srvNext_64Guest', 276, now(), 0);
+ALTER TABLE `cloud`.`s2s_customer_gateway` ADD COLUMN `ike_version` varchar(5) NOT NULL DEFAULT 'ike' COMMENT 'one of ike, ikev1, ikev2';
+ALTER TABLE `cloud`.`s2s_customer_gateway` ADD COLUMN `split_connections` int(1) NOT NULL DEFAULT 0;
```
Comment (DaanHoogland) [inline engine/schema/src/main/resources/META-INF/db/schema-41500to41510.sql:37]: are these two lines related to this change?

**Review by DaanHoogland — COMMENTED** [review]
2*q

File: engine/schema/src/main/resources/META-INF/db/schema-41500to41510.sql (hunk 7 above)
Comment (weizhouapache) [reply engine/schema/src/main/resources/META-INF/db/schema-41500to41510.sql:37]: 4.15.1.0 already contains these two lines. this PR just moves the lines to the end of the sql file.
it should be ok.

Comment (sureshanaparti) [reply engine/schema/src/main/resources/META-INF/db/schema-41500to41510.sql:37]: correct @weizhouapache

File: engine/schema/src/main/java/com/cloud/upgrade/GuestOsMapper.java (hunk 6 above)
Comment (sureshanaparti) [reply engine/schema/src/main/java/com/cloud/upgrade/GuestOsMapper.java:43]: > is the `display_name` the only field we'd want to update?

yes @DaanHoogland

**Review by borisstoyanov — CHANGES_REQUESTED** [review]
@sureshanaparti I've hit some sql exceptions around adding guestoss during upgrade. can you have a look pls.
[sql-upgrade-exceptions.txt](https://github.com/apache/cloudstack/files/7971183/sql-upgrade-exceptions.txt)

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test

**Review by DaanHoogland — APPROVED** [review]
code looks good, it will have to go through some upgrade paths to be proven, but :+1:

**Review by borisstoyanov — APPROVED** [review]
LGTM, @sureshanaparti can you see what's wrong with the test failures

**Comment by sureshanaparti** [comment]
> LGTM, @sureshanaparti can you see what's wrong with the test failures

which test failures @borisstoyanov

**Comment by DaanHoogland** [comment]
> > LGTM, @sureshanaparti can you see what's wrong with the test failures
> 
> which test failures @borisstoyanov

older test results, I had deleted. the newest are fine @sureshanaparti

File: engine/schema/src/main/java/com/cloud/upgrade/GuestOsMapper.java [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+import java.sql.SQLException;
+import java.util.List;
+
+import com.cloud.utils.Ternary;
+
+public class GuestOsMapper {
+
+    final static Logger LOG = Logger.getLogger(GuestOsMapper.class);
+
+    private static final String selectGuestOsSql =
```
Comment (rohityadavcloud) [inline engine/schema/src/main/java/com/cloud/upgrade/GuestOsMapper.java:34]: @sureshanaparti could we use Daos instead of manually created raw sql command?

**Comment by rohityadavcloud** [comment]
@sureshanaparti it appears this fixes only for specific upgrade paths, while it may fix the mapping for an existing env, say in future the mapping issue happens again - then we would need to use the utility again in the upgrade path?

**Comment by DaanHoogland** [comment]
@rohityadavcloud I think we will have to use this utility from here on in always. same for the annotations utility.

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan test keepEnv

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan test keepEnv

File: engine/schema/src/main/java/com/cloud/storage/GuestOSVO.java [hunk 9]
```diff
@@ -113,7 +113,7 @@ public Date getRemoved() {
     }
 
     @Override
-    public boolean isUserDefined() {
+    public boolean getIsUserDefined() {
```
Comment (DaanHoogland) [inline engine/schema/src/main/java/com/cloud/storage/GuestOSVO.java:116]: not a good rename

Comment (sureshanaparti) [reply engine/schema/src/main/java/com/cloud/storage/GuestOSVO.java:116]: agree, but i think it works if field name is _userDefined_

https://github.com/apache/cloudstack/blob/41d25d54a79c21877b4c493261ab6a60251b9f34/framework/db/src/main/java/com/cloud/utils/db/SearchBase.java#L484-L486

https://github.com/apache/cloudstack/blob/41d25d54a79c21877b4c493261ab6a60251b9f34/framework/db/src/main/java/com/cloud/utils/db/SqlGenerator.java#L83-L84

File: api/src/main/java/com/cloud/storage/GuestOS.java [hunk 10]
```diff
@@ -33,5 +33,5 @@
 
     Date getRemoved();
 
-    boolean isUserDefined();
+    boolean getIsUserDefined();
```
[the remark this answers is not in the document]
Comment (sureshanaparti) [reply api/src/main/java/com/cloud/storage/GuestOS.java:36]: field name has to be _<property-name>_, not _is+<property-name>_

[the remark this answers is not in the document]
Comment (DaanHoogland) [reply api/src/main/java/com/cloud/storage/GuestOS.java:36]: I see @sureshanaparti , In that case I think we should rename the field, not the accessor.

[the remark this answers is not in the document]
Comment (sureshanaparti) [reply api/src/main/java/com/cloud/storage/GuestOS.java:36]: correct, changed the accessor to be in line with GuestOSHypervisor. May be, have to fix this is Dao later.

https://github.com/apache/cloudstack/blob/850714c1fcce7ae68533eac522f7a1f9748641c0/api/src/main/java/com/cloud/storage/GuestOSHypervisor.java#L39

https://github.com/apache/cloudstack/blob/850714c1fcce7ae68533eac522f7a1f9748641c0/engine/schema/src/main/java/com/cloud/storage/GuestOSHypervisorVO.java#L128

**Review by borisstoyanov — APPROVED** [review]
LGTM, manually checked

File: api/src/main/java/com/cloud/storage/GuestOS.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (DaanHoogland) [reply api/src/main/java/com/cloud/storage/GuestOS.java:36]: ok, this is bad technmical debt, but for the fix now, ok, you convinced me (off-line)
